src/django_components/component.py

```diff
@@ -1022,101 +1085,173 @@ def _render_impl(
             # to see if given slot was filled, e.g.:
             # `{% if variable > 8 and component_vars.is_filled.header %}`
             is_filled = SlotIsFilled(slots_untyped)
-            self._render_stack[-1].is_filled = is_filled
-
-            # If any slot fills were defined within the template, we want to scope them
-            # to the CSS of the parent component. Thus we keep track of the parent component.
-            if context.get(_COMPONENT_SLOT_CTX_CONTEXT_KEY, None):
-                parent_comp_ctx: ComponentSlotContext = context[_COMPONENT_SLOT_CTX_CONTEXT_KEY]
-                parent_id = parent_comp_ctx.component_id
-            else:
-                parent_id = None
-
-            component_slot_ctx = ComponentSlotContext(
-                component_name=self.name,
-                component_id=render_id,
-                template_name=template.name,
-                fills=slots_untyped,
-                is_dynamic_component=getattr(self, "_is_dynamic_component", False),
-                # This field will be modified from within `SlotNodes.render()`:
-                # - The `default_slot` will be set to the first slot that has the `default` attribute set.
-                #   If multiple slots have the `default` attribute set, yet have different name, then
-                #   we will raise an error.
-                default_slot=None,
-            )
+            metadata.is_filled = is_filled
 
             with context.update(
                 {
                     # Private context fields
-                    _ROOT_CTX_CONTEXT_KEY: self.outer_context,
-                    _COMPONENT_SLOT_CTX_CONTEXT_KEY: component_slot_ctx,
-                    _REGISTRY_CONTEXT_KEY: self.registry,
+                    _COMPONENT_CONTEXT_KEY: render_id,
                     # NOTE: Public API for variables accessible from within a component's template
                     # See https://github.com/django-components/django-components/issues/280#issuecomment-2081180940
                     "component_vars": ComponentVars(
                         is_filled=is_filled,
                     ),
                 }
             ):
-                self.on_render_before(context, template)
+                # Make a "snapshot" of the context as it was at the time of the render call.
+                #
+                # Previously, we recursively called `Template.render()` as this point, but due to recursion
+                # this was limiting the number of nested components to only about 60 levels deep.
+                #
+                # Now, we make a flat copy, so that the context copy is static and doesn't change even if
+                # we leave the `with context.update` blocks.
+                #
+                # This makes it possible to render nested components with a queue, avoiding recursion limits.
+                context_snapshot = snapshot_context(context)
```

Comment:
This is where we previously called `Template.render(context)`, and this is where we now make only the copy or "snapshot" of the context.